### PR TITLE
WASM Update

### PR DESCRIPTION
**Pull request by neophob**

### CODE REVIEW

File: wasm/hal-wasm.go [hunk 1]
```diff
@@ -31,14 +31,15 @@ func (render *WASMHAL) updateKeyStateFrom(keyMap *map[uint32]bool) {
 
 func (render *WASMHAL) BlitPage(buffer [anotherworld.WIDTH * anotherworld.HEIGHT]anotherworld.Color, posX, posY int) {
 	// logger.Debug(">VID: BLITPAGE %d %d", posX, posY)
+	offset := 0
 	for y := 0; y < int(anotherworld.HEIGHT); y++ {
 		for x := 0; x < int(anotherworld.WIDTH); x++ {
-			offset := y*int(anotherworld.WIDTH) + x
```
Comment (neophob) [inline wasm/hal-wasm.go:36]: less mul's per blit

File: wasm/js-api.go [hunk 2]
```diff
@@ -65,6 +65,7 @@ func startGameFromPartWrapper(engine *Engine, inputs []js.Value) {
 		engine.app.LoadGamePart(startPartId)
 	}
 
+	engine.app.LoadGamePart(anotherworld.GAME_PART_ID_2)
```
Comment (neophob) [inline wasm/js-api.go:68]: load the intro part - like the SDL Version

File: wasm/js-canvas.go [hunk 3]
```diff
@@ -25,14 +25,21 @@ func GetCanvas(domElementId string) Canvas {
...
+//  pixels[3] = 255
 
 func (c Canvas) SetColor(color anotherworld.Color) {
-	c.fillStyle(getHexColorFrom(color))
+	c.context2d.Set("fillStyle", fmt.Sprintf("#%X%X%X", color.R, color.G, color.B))
 }
 
 func (c Canvas) DrawPoint(x, y int) {
-	c.fillRect(x, y, 1, 1)
+	c.context2d.Call("fillRect", x, y, 1, 1)
```
Comment (neophob) [inline wasm/js-canvas.go:42]: avoid function calls - saves W*H function calls per blit

File: wasm/js-canvas.go [hunk 4]
```diff
@@ -25,14 +25,21 @@ func GetCanvas(domElementId string) Canvas {
...
+//  var id = ctx.getImageData(0, 0, canvasWidth, canvasHeight)
+//  var pixels = id.data
+//  pixels[0] = 0
+//  pixels[1] = 255
+//  pixels[2] = 255
+//  pixels[3] = 255
 
 func (c Canvas) SetColor(color anotherworld.Color) {
-	c.fillStyle(getHexColorFrom(color))
+	c.context2d.Set("fillStyle", fmt.Sprintf("#%X%X%X", color.R, color.G, color.B))
```
Comment (neophob) [inline wasm/js-canvas.go:38]: avoid function calls - saves 2*W*H function calls per blit

File: wasm/js-api.go [hunk 5]
```diff
@@ -59,11 +59,8 @@ func startGameFromPartWrapper(engine *Engine, inputs []js.Value) {
 		startPartId += parseGamePartOffset(inputs[0])
 	}
 
-	nonDefaultPartLoadingIsNeeded := startPartId != defaultStartPart
```
Comment (neophob) [inline wasm/js-api.go:62]: this did not work as startPartId was set to defaultStartPart ;)

File: wasm/index.html [hunk 6]
```diff
@@ -8,22 +8,22 @@
...
-        height:200px;
+        width: 640px;
+        height: 400px;
         display: block;
         margin-left:auto;
         margin-right:auto;
         -webkit-box-shadow: 10px 10px 5px 0px rgba(0,0,0,0.75);
         -moz-box-shadow: 10px 10px 5px 0px rgba(0,0,0,0.75);
         box-shadow: 10px 10px 5px 0px rgba(0,0,0,0.75);
+        image-rendering: pixelated;
```
Comment (pfiaux) [inline wasm/index.html:19]: oh

File: wasm/js-canvas.go [hunk 7]
```diff
@@ -25,24 +25,27 @@ func GetCanvas(domElementId string) Canvas {
...
+//  var id = ctx.getImageData(0, 0, canvasWidth, canvasHeight)
+//  var pixels = id.data
+//  pixels[0] = 0
+//  pixels[1] = 255
+//  pixels[2] = 255
+//  pixels[3] = 255
 
 func (c Canvas) SetColor(color anotherworld.Color) {
-	c.fillStyle(getHexColorFrom(color))
+	c.context2d.Set("fillStyle", fmt.Sprintf("rgb(%d, %d, %d)", color.R, color.G, color.B))
```
Comment (pfiaux) [inline wasm/js-canvas.go:38]: do you think the sprintf here needs to be optimized? Not sure what it does under the hood.

Comment (neophob) [reply wasm/js-canvas.go:38]: looking at the performance its more about the fill Style call. I have an idea how to optimize in a later pr

File: wasm/hal-wasm.go [hunk 8]
```diff
@@ -31,14 +31,17 @@ func (render *WASMHAL) updateKeyStateFrom(keyMap *map[uint32]bool) {
 
 func (render *WASMHAL) BlitPage(buffer [anotherworld.WIDTH * anotherworld.HEIGHT]anotherworld.Color, posX, posY int) {
 	// logger.Debug(">VID: BLITPAGE %d %d", posX, posY)
+	render.canvas.SetColor(buffer[0])
+	render.lastSetColor = buffer[0]
+	offset := 0
```
Comment (pfiaux) [inline wasm/hal-wasm.go:36]: I found myself wondering, what offset? should we call it xYOffset or a similarly more descriptive variant

File: wasm/hal-wasm.go [hunk 9]
```diff
@@ -31,14 +31,17 @@ func (render *WASMHAL) updateKeyStateFrom(keyMap *map[uint32]bool) {
...
 	for y := 0; y < int(anotherworld.HEIGHT); y++ {
 		for x := 0; x < int(anotherworld.WIDTH); x++ {
-			offset := y*int(anotherworld.WIDTH) + x
 			if color := buffer[offset]; color != render.lastSetColor {
 				render.canvas.SetColor(color)
 				render.lastSetColor = color
 			}
 			render.canvas.DrawPoint(x+posX, y+posY)
```
Comment (pfiaux) [inline wasm/hal-wasm.go:43]: Would doing `absoluteY := y+posY` outside of the x loop help? less additions per blit or over optimization?

File: wasm/index.html [hunk 10]
```diff
@@ -8,22 +8,22 @@
...
       }
     </style>
   </head>
   <body>
     <!-- TODO use fancy webworkers for caching all the assets... -->
     <h1><img src="logo.png" alt="Gother World"></h1>
-    <canvas id="gotherworld"/>
+    <canvas id="gotherworld" width="320" height="200"/>
```
Comment (pfiaux) [inline wasm/index.html:26]: minor: something like `canvasElement.Set("width", 320)` and ... in js-canvas.go might also work if we want to leverage consts in go.

File: wasm/js-canvas.go [hunk 11]
```diff
@@ -25,24 +25,27 @@ func GetCanvas(domElementId string) Canvas {
...
 }
 
-func (c Canvas) fillRect(x, y, length, width int) {
-	c.context2d.Call("fillRect", x, y, length, width)
-}
-
-func getHexColorFrom(color anotherworld.Color) string {
-	return fmt.Sprintf("#%X%X%X", color.R, color.G, color.B)
+func (c Canvas) fillRect(x, y, width, height int) {
+	c.context2d.Call("fillRect", x, y, width, height)
```
Comment (pfiaux) [inline wasm/js-canvas.go:50]: if you're calling `c.context2d.Call("fillRect", x, y, 1, 1)` directly above is this function still needed?

File: wasm/js-canvas.go (hunk 7 above)
Comment (pfiaux) [reply wasm/js-canvas.go:38]: Sounds good, I also have some links in https://github.com/neophob/ganother-world/issues/30 I didn't get around to test

File: wasm/hal-wasm.go (hunk 9 above)
Comment (neophob) [reply wasm/hal-wasm.go:43]: add is much cheaper than a mul, should be fine imho unless perf meter says something else

File: wasm/hal-wasm.go (hunk 8 above)
Comment (neophob) [reply wasm/hal-wasm.go:36]: there is only one offset in the whole buffer, we could rename it to xYOffset but imho this does not improve the readability - its just an offset

File: wasm/index.html (hunk 10 above)
Comment (neophob) [reply wasm/index.html:26]: yes we could but the game itself is fixed at this resolution

**Pull request merged**